[CMake] Allow using a system-provided libsysprof-capture package

aperezdc opened the pull request:
#### 9cdef20e38c0e2ca0c5897b01c60c2a0754631bb
<pre>
[CMake] Allow using a system-provided libsysprof-capture package
<a href="https://bugs.webkit.org/show_bug.cgi?id=275321">https://bugs.webkit.org/show_bug.cgi?id=275321</a>

Reviewed by Michael Catanzaro.

Add a new USE_SYSTEM_SYSPROF_CAPTURE build option and a corresponding
CMake find-module. The module defines a SysProfCapture::SysProfCapture
target, the same name as already used for the copy of the library that
was already bundled under Source/ThirdParty/, therefore chosing between
the both does not need any other changes than using the find-module when
the CMake option is set.

* Source/CMakeLists.txt: Switch between the find-module for
sysprof-capture and the bundled copy depending on the value
of USE_SYSTEM_SYSPROF_CAPTURE.
* Source/cmake/FindSysProfCapture.cmake: Added.
* Source/cmake/OptionsGTK.cmake: Add USE_SYSTEM_SYSPROF_CAPTURE build
option, move the USE_SYSPROF_CAPTURE option definition here.
* Source/cmake/OptionsWPE.cmake: Ditto.
* Source/cmake/WebKitFeatures.cmake: Remove the USE_SYSPROF_CAPTURE
build option definition, it should be port-specific.

Canonical link: <a href="https://commits.webkit.org/282340@main">https://commits.webkit.org/282340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f937212a2db6eb8de0b8afc5cf58641e21a04120

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42247 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66912 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/13495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13779 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/13495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65960 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/54475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/35955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12371 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/56001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/12138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68607 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/62134 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6837 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6869 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/54535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/5709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83897 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9472 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/14766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/WebKitFeatures.cmake:
```diff
@@ -267,6 +267,7 @@ macro(WEBKIT_OPTION_BEGIN)
     WEBKIT_OPTION_DEFINE(USE_SKIA "Whether to use Skia instead of Cairo." PRIVATE OFF)
     WEBKIT_OPTION_DEFINE(USE_SYSPROF_CAPTURE "Whether to use libsysprof-capture for tracing." PRIVATE OFF)
     WEBKIT_OPTION_DEFINE(USE_SYSTEM_MALLOC "Toggle system allocator instead of WebKit's custom allocator" PRIVATE ${USE_SYSTEM_MALLOC_DEFAULT})
+    WEBKIT_OPTION_DEFINE(USE_SYSTEM_SYSPROF_CAPTURE "Whether to use a system-provided libsysprof-capture" PRIVATE OFF)
```
Oops, both this and the preexisting `USE_SYSPROF_CAPTURE` option should go in OptionsGTK.cmake and OptionsWPE.cmake instead, since they're not relevant to all platforms. Same for `USE_SKIA`. The idea is that options listed here should be relevant to more than just WPE/GTK.

E.g. `USE_AVIF`, `USE_JPEGXL`, `USE_LCMS` do not depend on GLib and are theoretically relevant to any port other than the Apple ports, so it's fine to list those here. But the sysprof stuff is exclusively useful for WPE/GTK because it depends on GLib.

One more thing: this should be enabled by default, because using the system version is preferable to using the bundled version.

aperezdc replied:
> Oops, both this and the preexisting `USE_SYSPROF_CAPTURE` option should go in OptionsGTK.cmake and OptionsWPE.cmake instead, since they're not relevant to all platforms. Same for `USE_SKIA`. The idea is that options listed here should be relevant to more than just WPE/GTK.

I'll move the Sysprof options to `Options{WPE,GTK}.cmake`.

OTOH, `USE_SKIA` is *already* cross-port: PlayStation has the plumbing to use it, and WinCairo is getting there as well. Skia doesn't depend on GLib either.

> One more thing: this should be enabled by default, because using the system version is preferable to using the bundled version.

No. I want the bundled one by default, otherwise packagers are more likely to disable the support completely via `USE_SYSPROF_CAPTURE=OFF` instead of setting `USE_SYSTEM_SYSPROF_CAPTURE=OFF`. Trust me, I have packaged software for years (and still do), and I know well how they think. As long as we provide an option to let them pick the system package easily, they won't be angry at us, either.

mcatanzaro replied:
The end result will be distros using the bundled version, since that is default. If this is what you want and the system option is only intended to be an alternative, then what you have now is ideal. But if you want distros to actually use the system sysprof, then system sysprof should be default.

Our standard pattern would work fine here:

```
if (NOT SysProfCapture_FOUND)
    message(FATAL_ERROR "sysprof-capture is required for USE_SYSTEM_SYSPROF_CAPTURE")
endif ()
```

Packagers won't even see `USE_SYSPROF_CAPTURE` unless they go looking for it. They _will_ see `USE_SYSTEM_SYSPROF_CAPTURE` because it'll be at the bottom of the failed build log. So the path of least resistance will not lead toward disabled sysprof-capture.

The message could be reworded for further clarity; currently all our error messages assume some level of familiarity with CMake options. E.g. "sysprof-capture dependency is required because USE_SYSTEM_SYSPROF_CAPTURE is enabled; you may disable with -DUSE_SYSTEM_SYSPROF=OFF"

> OTOH, USE_SKIA is already cross-port: PlayStation has the plumbing to use it, and WinCairo is getting there as well. Skia doesn't depend on GLib either.

Fair point.

aperezdc replied:
@mcatanzaro Updated, I still needed to keep usage of the bundled copy while we do not have `libsysprof-capture-4` included in the SDKs. I have filed [bug #277627](https://bugs.webkit.org/show_bug.cgi?id=277627) to track this.

mcatanzaro commented on Source/CMakeLists.txt:
```diff
@@ -35,7 +35,15 @@ if (USE_SKIA)
 endif ()
 
 if (USE_SYSPROF_CAPTURE)
```
Unimportant: you could alternatively write:

```
if (USE_SYSTEM_SYSPROF_CAPTURE)
    # ...
elseif (USE_SYSPROF_CAPTURE)
    # ...
endif ()
```

and then you have one fewer level of indentation, if you prefer.

aperezdc replied:
Doing this would require also setting a dependency between both options, which is probably a good thing to add anyway to ensure that `USE_SYSTEM_SYSPROF_CAPTURE` will always be disabled when `USE_SYSPROF_CAPTURE` is disabled as well.

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -51,6 +58,8 @@ WEBKIT_OPTION_DEFINE(USE_LIBDRM "Whether to enable usage of libdrm." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBSECRET "Whether to enable the persistent credential storage using libsecret." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_SOUP2 "Whether to enable usage of Soup 2 instead of Soup 3." PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_SYSPROF_CAPTURE "Whether to use libsysprof-capture for tracing." PRIVATE ON)
+WEBKIT_OPTION_DEFINE(USE_SYSTEM_SYSPROF_CAPTURE "Whether to use a system-provided libsysprof-capture" PRIVATE ${USE_SYSTEM_SYSPROF_CAPTURE_DEFAULT})
```
Move these lower. These are the public options. The list of private options is below.

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -101,6 +106,8 @@ WEBKIT_OPTION_DEFINE(USE_GBM "Whether to enable usage of GBM." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBDRM "Whether to enable usage of libdrm." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_QT6 "Whether to use Qt6 instead of Qt5" PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(USE_SYSPROF_CAPTURE "Whether to use libsysprof-capture for tracing." PRIVATE ON)
+WEBKIT_OPTION_DEFINE(USE_SYSTEM_SYSPROF_CAPTURE "Whether to use a system-provided libsysprof-capture" PRIVATE ${USE_SYSTEM_SYSPROF_CAPTURE_DEFAULT})
```
Here the list of private options is _immediately_ below.

aperezdc committed:
[CMake] Allow using a system-provided libsysprof-capture package
https://bugs.webkit.org/show_bug.cgi?id=275321

Reviewed by Michael Catanzaro.

Add a new USE_SYSTEM_SYSPROF_CAPTURE build option and a corresponding
CMake find-module. The module defines a SysProfCapture::SysProfCapture
target, the same name as already used for the copy of the library that
was already bundled under Source/ThirdParty/, therefore chosing between
the both does not need any other changes than using the find-module when
the CMake option is set.

* Source/CMakeLists.txt: Switch between the find-module for
sysprof-capture and the bundled copy depending on the value
of USE_SYSTEM_SYSPROF_CAPTURE.
* Source/cmake/FindSysProfCapture.cmake: Added.
* Source/cmake/OptionsGTK.cmake: Add USE_SYSTEM_SYSPROF_CAPTURE build
option, move the USE_SYSPROF_CAPTURE option definition here.
* Source/cmake/OptionsWPE.cmake: Ditto.
* Source/cmake/WebKitFeatures.cmake: Remove the USE_SYSPROF_CAPTURE
build option definition, it should be port-specific.

Canonical link: https://commits.webkit.org/282340@main
